lib/veri.ml

```diff
@@ -0,0 +1,237 @@
+open Core_kernel.Std
+open Bap.Std
+open Bap_traces.Std
+open Trace
+
+module SM = Monad.State
+open SM
+
+type point = {
+  code : Chunk.t;
+  side : event list
+}
+
+type 'a u = 'a Bil.Result.u
+type 'a r = 'a Bil.Result.r
+type 'a e = (event option, 'a) Monad.State.t
+
+let value = Bil.Result.value
+let make_point code = {code; side = [];}
+let add_side p ev = {p with side = ev :: p.side}
+let do_nothing () = SM.return ()
+
+let create_move_event tag cell' data' =  
+  Value.create tag Move.({cell = cell'; data = data';})   
+
+let find tag evs cond =
+  let open Option in
+  List.find evs ~f:(fun ev -> match Value.get tag ev with
+      | None -> false
+      | Some mv -> cond mv) >>| fun ev -> Value.get_exn tag ev
+
+let create_mem_store = create_move_event Event.memory_store
+let create_mem_load  = create_move_event Event.memory_load
+let create_reg_read  = create_move_event Event.register_read
+let create_reg_write = create_move_event Event.register_write
+let find_reg_read = find Event.register_read
+let find_mem_load = find Event.memory_load
+
+let is_imm_var var = 
+  let open Type in
+  match Var.typ var with
+  | Imm _ -> true
+  | _ -> false
+
+let imm_result v = 
+  let open Bil in
+  match v with
+  | Imm x -> Some x
+  | _ -> None
+  
+module Set = Value.Set
+
+class context report trace = object(self:'s)
+  inherit Veri_traci.context trace as super
+  val report : Veri_report.t = report
+  val events : Set.t = Set.empty
+  val selves = None
+  val current = `Fst
+  val descr : string option = None
+
+  method! update var res = 
+    match selves with
+    | None -> super#update var res
+    | Some (main, shadow) ->
```

Comment:
You finally got 3 selves, instead of two. The idea was to have only two. One that is current (accessible via `self` instance variable), and another, that is currently sleeping. When you change selves, you just pick that one that was sleeping and return it, and put previously active sleep into sleep. 

Your three selves approach will not work, because you need to rewrite every method to dispatch between two selves, and the set of methods is open.